pkg/registry/apis/query/query.go

```diff
@@ -150,8 +246,67 @@ func (b *QueryAPIBuilder) executeConcurrentQueries(ctx context.Context, requests
 	return resp, nil
 }
 
-// NOTE the upstream queries have already been executed
-// https://github.com/grafana/grafana/blob/v10.2.3/pkg/services/query/query.go#L242
-func (b *QueryAPIBuilder) handleExpressions(ctx context.Context, qdr *backend.QueryDataResponse, expressions []v0alpha1.GenericDataQuery) (*backend.QueryDataResponse, error) {
-	return qdr, fmt.Errorf("expressions are not implemented yet")
+// Unlike the implementation in expr/node.go, all datasource queries have been processed first
+func (b *QueryAPIBuilder) handleExpressions(ctx context.Context, req parsedRequestInfo, data *backend.QueryDataResponse) (qdr *backend.QueryDataResponse, err error) {
+	start := time.Now()
+	ctx, span := b.tracer.Start(ctx, "SSE.handleExpressions")
+	defer func() {
+		var respStatus string
+		switch {
+		case err == nil:
+			respStatus = "success"
+		default:
+			respStatus = "failure"
+		}
+		duration := float64(time.Since(start).Nanoseconds()) / float64(time.Millisecond)
+		b.metrics.expressionsQuerySummary.WithLabelValues(respStatus).Observe(duration)
+
+		span.End()
+	}()
+
+	qdr = data
+	if qdr == nil {
+		qdr = &backend.QueryDataResponse{}
+	}
+	now := start // <<< this should come from the original query parser
+	vars := make(mathexp.Vars)
+	for _, expression := range req.Expressions {
+		// Setup the variables
+		for _, refId := range expression.Command.NeedsVars() {
+			_, ok := vars[refId]
+			if !ok {
+				dr, ok := qdr.Responses[refId]
+				if ok {
+					allowLongFrames := false // TODO -- depends on input type and only if SQL?
+					_, res, err := b.converter.Convert(ctx, req.RefIDTypes[refId], dr.Frames, allowLongFrames)
+					if err != nil {
+						res.Error = err
+					}
+					vars[expression.RefID] = res
```

Comment:
what's the difference between `expression.RefID` and `refId`?

Reply:
a mistake 🤦🏻 -- fixed.  expression.RefID is the expressions ID, but "refId" is the id for the result it requires.  The map should be full of the results id